### PR TITLE
Consolidate and expand discussion of supported indices

### PR DESCRIPTION
**Pull request by mbauman**
This greatly expands the discussion of supported index types in a more rigorous fashion. Adds much more discussion about CartesianIndex and arrays thereof, along with some new examples.

Is it kosher to link out to a blog post from the manual?

### CODE REVIEW

File: doc/src/manual/arrays.md [hunk 1]
```diff
@@ -230,13 +230,11 @@ The general syntax for indexing into an n-dimensional array A is:
...
-where each `I_k` may be:
-
-1. A scalar integer
-2. A `Range` of the form `a:b`, or `a:b:c`
-3. A `:` or `Colon()` to select entire dimensions
-4. An arbitrary integer array, including the empty array `[]`
-5. A boolean array to select a vector of elements at its `true` indices
+where each `I_k` may be a scalar integer, an array of integers, or any other
+[supported index](@ref man-supported-index-types). This includes
+[`Colon`](@ref) (`:`) to select all indices within the entire dimension,
```
Comment (kshyatt) [inline doc/src/manual/arrays.md]: Will this ref work? Did you try it in HTML?

File: doc/src/manual/arrays.md [hunk 2]
```diff
@@ -230,13 +230,11 @@ The general syntax for indexing into an n-dimensional array A is:
...
+ranges of the form `a:c` or `a:b:c` to select congiguous or strided
+subsections, and arrays of booleans to select elements at their `true` indices.
```
Comment (kshyatt) [inline doc/src/manual/arrays.md]: A doctest example would be cool here.

**Review by kshyatt — APPROVED** [review]
Assuming that `Colon` link works this is awesome!

File: doc/src/manual/arrays.md (hunk 1 above)
Comment (mbauman) [inline doc/src/manual/arrays.md]: Yup! It links to http://docs.julialang.org/en/latest/stdlib/arrays.html#Base.Colon

**Comment by mbauman** [comment]
Thanks for the review!

File: doc/src/manual/arrays.md [hunk 3]
```diff
@@ -230,13 +230,11 @@ The general syntax for indexing into an n-dimensional array A is:
...
-
-1. A scalar integer
-2. A `Range` of the form `a:b`, or `a:b:c`
-3. A `:` or `Colon()` to select entire dimensions
-4. An arbitrary integer array, including the empty array `[]`
-5. A boolean array to select a vector of elements at its `true` indices
+where each `I_k` may be a scalar integer, an array of integers, or any other
+[supported index](@ref man-supported-index-types). This includes
+[`Colon`](@ref) (`:`) to select all indices within the entire dimension,
+ranges of the form `a:c` or `a:b:c` to select congiguous or strided
```
Comment (tkelman) [inline doc/src/manual/arrays.md]: contiguous is misspelled

File: doc/src/manual/arrays.md [hunk 4]
```diff
@@ -317,23 +296,22 @@ The general syntax for assigning values in an n-dimensional array A is:
...
-
-1. A scalar integer
-2. A `Range` of the form `a:b`, or `a:b:c`
-3. A `:` or `Colon()` to select entire dimensions
-4. An arbitrary integer array, including the empty array `[]`
-5. A boolean array to select elements at its `true` indices
+where each `I_k` may be a scalar integer, an array of integers, or any other
+[supported index](@ref man-supported-index-types). This includes
+[`Colon`](@ref) (`:`) to select all indices within the entire dimension,
+ranges of the form `a:c` or `a:b:c` to select congiguous or strided
```
Comment (tkelman) [inline doc/src/manual/arrays.md]: and here

File: doc/src/manual/arrays.md [hunk 5]
```diff
@@ -358,6 +336,141 @@ julia> x
...
+  6  22
+ 11  27
+ 16  32
+```
+
+!!! warning
+
+    `CartesianIndex` and arrays of `CartesianIndex` are not compatible with the
+    `end` keyword to represent the last index of a dimension. Do not use `end`
+    in indexing expressions that may contain either.
```
Comment (tkelman) [inline doc/src/manual/arrays.md]: this sentence feels incomplete

File: doc/src/manual/arrays.md (hunk 3 above)
Comment (mbauman) [inline doc/src/manual/arrays.md]: oops.

File: doc/src/manual/arrays.md (hunk 5 above)
Comment (mbauman) [inline doc/src/manual/arrays.md]: I think it's technically correct to use either as a pronoun here, but I hear you.  How about "… may contain either kind of index." ?

Comment (tkelman) [inline doc/src/manual/arrays.md]: That would be an improvement, and make the subject more obvious. Would be most explicit but a bit repetitive to just say what's being referred to.

Comment (mbauman) [inline doc/src/manual/arrays.md]: Folded into #20223.

**Pull request merged**